File: python-algorithm/algo_08_graph/graph_3_toposort/L210-m.cpp
//
// Created by LiHeng on 2020/7/12.
//

//DFS实现拓扑排序

#include <iostream>
#include <vector>
using  namespace std;
class Solution {
public:
    vector<int> res;
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> flag(numCourses, 0);
        vector<vector<int>> adjacency(numCourses, vector<int>());
        for(auto vec: prerequisites)
            adjacency[vec[1]].push_back(vec[0]);
        for(int i=0; i<numCourses; ++i)
            if (dfs(i, flag, adjacency))
                return vector<int>();
        reverse(res.begin(), res.end()); //为什么要反序，思考下
        return res;
    }

    //返回true代表有向图有环，false代表无环
    bool dfs(int index, vector<int>& flag, vector<vector<int>>& adjacency){
        if(flag[index] == -1) return false;
        if(flag[index] == 1) return true;
        flag[index] = 1;
        for(auto j:adjacency[index]) //adjacency[index]里面存放的是学完index课程后可以学的其他课程
            if (dfs(j, flag, adjacency))
                return true;
        flag[index] = -1;
        res.push_back(index);
        return false;
    }
};

//可以用案例 3 [[0,1],[0,2],[1,2]] 理解过程



//BFS实现的拓扑排序
#include <queue>
class Solution1 {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> res;
        vector<int> indegrees(numCourses, 0); //入度记录
        vector<vector<int>> adjacency(numCourses, vector<int>());
        for(auto vec: prerequisites){
            indegrees[vec[0]]++;
            adjacency[vec[1]].push_back(vec[0]);
        }
        queue<int> q1;
        for(int i=0; i<numCourses; i++)
            if(indegrees[i] == 0)
                q1.push(i);
        while(!q1.empty()){
            int pre = q1.front();
            q1.pop();
            res.push_back(pre);
            numCourses -= 1;
            for(auto cur:adjacency[pre]){
                indegrees[cur]--;
                if(indegrees[cur] == 0)
                    q1.push(cur);
            }
        }
        return numCourses == 0?res:vector<int>();
    }
};

File: python-algorithm/algo_08_graph/graph_3_toposort/L210-m.py
#--coding:utf-8--


# 1 拓扑排序
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        res = [] # res存储的就是能完成的顺序
        degree = [0 for _ in range(numCourses)]
        graph = [[] for _ in range(numCourses)]
        for p in prerequisites:
            graph[p[1]].append(p[0])
            degree[p[0]] += 1  # p[0]点入度加1
        q = collections.deque()
        # 将入度为0的节点全部加入队列
        for i in range(numCourses):
            if degree[i] == 0:
                q.append(i)
        while q:
            t = q.popleft()
            res.append(t)
            # 删除所有与当前节点相连的边
            for i in graph[t]:
                degree[i] -= 1
                # 度变成0的节点加入队列
                if degree[i] == 0:
                    q.append(i)
        if len(res) == numCourses:
            return res
        return []


# 2 DFS
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph = [[] for i in range(numCourses)]
        for connect in prerequisites:
            graph[connect[1]].append(connect[0])
        visited = [0 for i in range(numCourses)]
        res = []
        for i in range(numCourses):
            if self.dfs(graph, i, visited, res):
                return []
        return res[::-1] # 为什么要反序，思考下!

    def dfs(self, graph, idx, visited, res):
        if visited[idx] == -1:  # 提前剪枝
            return False
        if visited[idx] == 1:
            return True
        visited[idx] = 1
        for nex in graph[idx]:
            if self.dfs(graph, nex, visited, res):
                return True
        visited[idx] = -1  # 说明以index为起始点的遍历无环
        res.append(idx)
        return False